File: Terminus-Edition-2/aws-lambda-event-source-mapping-recovery/environment/src/simulator.py
from __future__ import annotations
import argparse
import json
from pathlib import Path
from typing import Any
from .jsonio import load_json, save_json
from .paths import config_path, ledger_path, dlq_path
from .mapping import active_mapping
from .iam_simulator import required_sqs_decisions, has_broad_sqs_grant, has_log_permissions, decide
from .state_store import load_ledger, load_dlq, append_dlq_entry
from . import handler_bridge

def load_runtime() -> tuple[dict[str, Any], dict[str, Any], dict[str, Any]]:
    return load_json(config_path("event_source_mapping.json")), load_json(config_path("queues.json")), load_json(config_path("lambda_role_policy.json"))

def mapping_probe() -> dict[str, Any]:
    mapping, queues, _policy = load_runtime()
    return active_mapping(mapping, queues)

def iam_probe() -> dict[str, Any]:
    _mapping, queues, policy = load_runtime()
    queue_arn = queues["expected_active_queue_arn"]
    return {"queue_arn": queue_arn, "decisions": required_sqs_decisions(policy, queue_arn), "old_queue_receive": decide(policy, "sqs:ReceiveMessage", queues["old_queue_arn"]), "has_broad_sqs_grant": has_broad_sqs_grant(policy), "has_log_permissions": has_log_permissions(policy)}

def _event_records(messages: list[dict[str, Any]]) -> list[dict[str, Any]]:
    records = []
    for msg in messages:
        cloned = dict(msg)
        attrs = dict(cloned.get("attributes", {}))
        attrs["ApproximateReceiveCount"] = str(int(attrs.get("ApproximateReceiveCount", "0")) + 1)
        cloned["attributes"] = attrs
        records.append(cloned)
    return records

def _message_reason(message: dict[str, Any]) -> str:
    try:
        body = json.loads(message.get("body", ""))
        if body.get("poison"):
            return body.get("failure_reason", "POISON_MESSAGE")
        return "PROCESSING_FAILED"
    except Exception:
        return "MALFORMED_JSON"

def simulate_batch(batch_file: str | Path, cycles: int = 1) -> dict[str, Any]:
    mapping, queues, policy = load_runtime()
    map_info = active_mapping(mapping, queues)
    batch = load_json(batch_file)
    redrive = load_json(config_path("redrive_policy.json"))
    queue_arn = batch.get("queue_arn", queues["expected_active_queue_arn"])
    result: dict[str, Any] = {"mapping": map_info, "queue_arn": queue_arn, "cycles": [], "access_denied": False, "access_denied_actions": [], "delivered_message_ids": [], "deleted_message_ids": [], "failed_message_ids": [], "receive_counts": {}, "dlq_message_ids": []}
    required = required_sqs_decisions(policy, queues["expected_active_queue_arn"])
    denied = [a for a, d in required.items() if d != "allowed"]
    if denied:
        result["access_denied"] = True
        result["access_denied_actions"] = denied
        result["ledger_entries"] = load_ledger(ledger_path())
        result["dlq_entries"] = load_dlq(dlq_path())
        return result
    if not map_info["active"] or queue_arn != queues["expected_active_queue_arn"]:
        result["ledger_entries"] = load_ledger(ledger_path())
        result["dlq_entries"] = load_dlq(dlq_path())
        return result
    messages = [dict(m) for m in batch.get("messages", [])]
    max_receive = int(redrive.get("max_receive_count", 3))
    deleted: set[str] = set()
    dlqed: set[str] = set()
    for _cycle in range(cycles):
        available = [m for m in messages if m.get("messageId") not in deleted and m.get("messageId") not in dlqed]
        if not available:
            result["cycles"].append({"delivered": [], "failed": [], "deleted": []})
            continue
        delivered = available[:int(mapping.get("batch_size", 10))]
        records = _event_records(delivered)
        for rec, original in zip(records, delivered):
            original.setdefault("attributes", {})["ApproximateReceiveCount"] = rec["attributes"]["ApproximateReceiveCount"]
            result["receive_counts"][original["messageId"]] = int(original["attributes"]["ApproximateReceiveCount"])
        response = handler_bridge.handle_batch({"Records": records})
        failures = response.get("batchItemFailures", []) if isinstance(response, dict) else []
        failed_ids = {item.get("itemIdentifier") for item in failures if isinstance(item, dict)}
        cycle_deleted = []
        cycle_failed = []
        for message in delivered:
            mid = message["messageId"]
            result["delivered_message_ids"].append(mid)
            if mid in failed_ids:
                result["failed_message_ids"].append(mid)
                cycle_failed.append(mid)
                receive_count = int(message.get("attributes", {}).get("ApproximateReceiveCount", "0"))
                if receive_count >= max_receive:
                    body_event_id = None
                    try:
                        body_event_id = json.loads(message.get("body", "{}")).get("business_event_id")
                    except Exception:
                        pass
                    append_dlq_entry(dlq_path(), {"message_id": f"dlq-{mid}", "original_message_id": mid, "business_event_id": body_event_id, "source_queue_arn": queue_arn, "failure_reason": _message_reason(message), "receive_count": receive_count})
                    dlqed.add(mid)
                    result["dlq_message_ids"].append(mid)
            else:
                deleted.add(mid)
                result["deleted_message_ids"].append(mid)
                cycle_deleted.append(mid)
        result["cycles"].append({"delivered": [m["messageId"] for m in delivered], "failed": cycle_failed, "deleted": cycle_deleted})
    result["ledger_entries"] = load_ledger(ledger_path())
    result["dlq_entries"] = load_dlq(dlq_path())
    return result

def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--scenario", choices=["mapping", "iam", "batch"], required=True)
    parser.add_argument("--batch", default="")
    parser.add_argument("--cycles", type=int, default=1)
    parser.add_argument("--result", default="")
    args = parser.parse_args(argv)
    if args.scenario == "mapping":
        result = mapping_probe()
    elif args.scenario == "iam":
        result = iam_probe()
    else:
        if not args.batch:
            parser.error("--batch is required for batch scenario")
        result = simulate_batch(args.batch, cycles=args.cycles)
    if args.result:
        save_json(args.result, result)
    print(json.dumps(result, indent=2, sort_keys=True))
    return 0

if __name__ == "__main__":
    raise SystemExit(main())

File: Terminus-Edition-2/aws-lambda-event-source-mapping-recovery/environment/simulator/simulator.go
package main

import (
	"encoding/json"
	"fmt"
	"strconv"
)

func stringValue(value any) string {
	switch typed := value.(type) {
	case string:
		return typed
	case nil:
		return ""
	default:
		return fmt.Sprint(typed)
	}
}

func intValue(value any, fallback int) int {
	switch typed := value.(type) {
	case float64:
		return int(typed)
	case float32:
		return int(typed)
	case int:
		return typed
	case int64:
		return int(typed)
	case json.Number:
		parsed, err := typed.Int64()
		if err == nil {
			return int(parsed)
		}
	case string:
		parsed, err := strconv.Atoi(typed)
		if err == nil {
			return parsed
		}
	}
	return fallback
}

func cloneMap(input map[string]any) map[string]any {
	data, _ := json.Marshal(input)
	var output map[string]any
	_ = json.Unmarshal(data, &output)
	return output
}

func loadRuntime() (map[string]any, map[string]any, map[string]any, error) {
	mapping := map[string]any{}
	queues := map[string]any{}
	policy := map[string]any{}
	if err := loadJSON(configPath("event_source_mapping.json"), &mapping); err != nil {
		return nil, nil, nil, err
	}
	if err := loadJSON(configPath("queues.json"), &queues); err != nil {
		return nil, nil, nil, err
	}
	if err := loadJSON(configPath("lambda_role_policy.json"), &policy); err != nil {
		return nil, nil, nil, err
	}
	return mapping, queues, policy, nil
}

func mappingProbe() (map[string]any, error) {
	mapping, queues, _, err := loadRuntime()
	if err != nil {
		return nil, err
	}
	return activeMapping(mapping, queues), nil
}

func iamProbe() (map[string]any, error) {
	_, queues, policy, err := loadRuntime()
	if err != nil {
		return nil, err
	}
	queueARN := stringValue(queues["expected_active_queue_arn"])
	return map[string]any{
		"queue_arn":           queueARN,
		"decisions":           requiredSQSDecisions(policy, queueARN),
		"old_queue_receive":   decide(policy, "sqs:ReceiveMessage", stringValue(queues["old_queue_arn"])),
		"has_broad_sqs_grant": hasBroadSQSGrant(policy),
		"has_log_permissions": hasLogPermissions(policy),
	}, nil
}

func eventRecords(messages []map[string]any) []map[string]any {
	records := make([]map[string]any, 0, len(messages))
	for _, message := range messages {
		cloned := cloneMap(message)
		attrs, _ := cloned["attributes"].(map[string]any)
		if attrs == nil {
			attrs = map[string]any{}
		}
		count := intValue(attrs["ApproximateReceiveCount"], 0) + 1
		attrs["ApproximateReceiveCount"] = strconv.Itoa(count)
		cloned["attributes"] = attrs
		records = append(records, cloned)
	}
	return records
}

func messageReason(message map[string]any) string {
	bodyText := stringValue(message["body"])
	body := map[string]any{}
	if err := json.Unmarshal([]byte(bodyText), &body); err != nil {
		return "MALFORMED_JSON"
	}
	poison, _ := body["poison"].(bool)
	if poison {
		if reason := stringValue(body["failure_reason"]); reason != "" {
			return reason
		}
		return "POISON_MESSAGE"
	}
	return "PROCESSING_FAILED"
}

func messageSlice(value any) []map[string]any {
	raw, _ := value.([]any)
	messages := make([]map[string]any, 0, len(raw))
	for _, item := range raw {
		if message, ok := item.(map[string]any); ok {
			messages = append(messages, cloneMap(message))
		}
	}
	return messages
}

func simulateBatch(batchFile string, cycles int) (map[string]any, error) {
	mapping, queues, policy, err := loadRuntime()
	if err != nil {
		return nil, err
	}
	mapInfo := activeMapping(mapping, queues)
	batch := map[string]any{}
	if err := loadJSON(batchFile, &batch); err != nil {
		return nil, err
	}
	redrive := map[string]any{}
	if err := loadJSON(configPath("redrive_policy.json"), &redrive); err != nil {
		return nil, err
	}
	queueARN := stringValue(batch["queue_arn"])
	if queueARN == "" {
		queueARN = stringValue(queues["expected_active_queue_arn"])
	}
	result := map[string]any{
		"mapping":               mapInfo,
		"queue_arn":             queueARN,
		"cycles":                []any{},
		"access_denied":         false,
		"access_denied_actions": []any{},
		"delivered_message_ids": []any{},
		"deleted_message_ids":   []any{},
		"failed_message_ids":    []any{},
		"receive_counts":        map[string]any{},
		"dlq_message_ids":       []any{},
	}

	expectedARN := stringValue(queues["expected_active_queue_arn"])
	required := requiredSQSDecisions(policy, expectedARN)
	denied := []any{}
	for _, action := range requiredSQSActions {
		if stringValue(required[action]) != "allowed" {
			denied = append(denied, action)
		}
	}
	if len(denied) > 0 {
		ledger, loadErr := loadLedger()
		if loadErr != nil {
			return nil, loadErr
		}
		dlq, loadErr := loadDLQ()
		if loadErr != nil {
			return nil, loadErr
		}
		result["access_denied"] = true
		result["access_denied_actions"] = denied
		result["ledger_entries"] = ledger
		result["dlq_entries"] = dlq
		return result, nil
	}

	active, _ := mapInfo["active"].(bool)
	if !active || queueARN != expectedARN {
		ledger, loadErr := loadLedger()
		if loadErr != nil {
			return nil, loadErr
		}
		dlq, loadErr := loadDLQ()
		if loadErr != nil {
			return nil, loadErr
		}
		result["ledger_entries"] = ledger
		result["dlq_entries"] = dlq
		return result, nil
	}

	messages := messageSlice(batch["messages"])
	maxReceive := intValue(redrive["max_receive_count"], 3)
	deleted := map[string]bool{}
	dlqed := map[string]bool{}
	currentDLQ, err := loadDLQ()
	if err != nil {
		return nil, err
	}
	for _, entry := range currentDLQ {
		if original := stringValue(entry["original_message_id"]); original != "" {
			dlqed[original] = true
		}
	}

	for cycle := 0; cycle < cycles; cycle++ {
		available := make([]map[string]any, 0, len(messages))
		for _, message := range messages {
			mid := stringValue(message["messageId"])
			if !deleted[mid] && !dlqed[mid] {
				available = append(available, message)
			}
		}
		cyclesOut := result["cycles"].([]any)
		if len(available) == 0 {
			result["cycles"] = append(cyclesOut, map[string]any{
				"delivered": []any{}, "failed": []any{}, "deleted": []any{},
			})
			continue
		}
		batchSize := intValue(mapping["batch_size"], 10)
		if batchSize > len(available) {
			batchSize = len(available)
		}
		delivered := available[:batchSize]
		records := eventRecords(delivered)
		receiveCounts := result["receive_counts"].(map[string]any)
		for index, record := range records {
			original := delivered[index]
			attrs := record["attributes"].(map[string]any)
			originalAttrs, _ := original["attributes"].(map[string]any)
			if originalAttrs == nil {
				originalAttrs = map[string]any{}
			}
			originalAttrs["ApproximateReceiveCount"] = attrs["ApproximateReceiveCount"]
			original["attributes"] = originalAttrs
			receiveCounts[stringValue(original["messageId"])] = intValue(originalAttrs["ApproximateReceiveCount"], 0)
		}

		rawRecords := make([]any, 0, len(records))
		for _, record := range records {
			rawRecords = append(rawRecords, record)
		}
		response, err := handleBatch(map[string]any{"Records": rawRecords})
		if err != nil {
			return nil, err
		}
		failedIDs := map[string]bool{}
		if rawFailures, ok := response["batchItemFailures"].([]any); ok {
			for _, rawFailure := range rawFailures {
				failure, ok := rawFailure.(map[string]any)
				if !ok {
					continue
				}
				failedIDs[stringValue(failure["itemIdentifier"])] = true
			}
		}

		cycleDeleted := []any{}
		cycleFailed := []any{}
		deliveredIDs := make([]any, 0, len(delivered))
		for _, message := range delivered {
			mid := stringValue(message["messageId"])
			deliveredIDs = append(deliveredIDs, mid)
			result["delivered_message_ids"] = append(result["delivered_message_ids"].([]any), mid)
			if failedIDs[mid] {
				result["failed_message_ids"] = append(result["failed_message_ids"].([]any), mid)
				cycleFailed = append(cycleFailed, mid)
				attrs, _ := message["attributes"].(map[string]any)
				receiveCount := intValue(attrs["ApproximateReceiveCount"], 0)
				if receiveCount >= maxReceive {
					var eventID any
					body := map[string]any{}
					if err := json.Unmarshal([]byte(stringValue(message["body"])), &body); err == nil {
						eventID = body["business_event_id"]
					}
					if err := appendDLQEntry(map[string]any{
						"message_id":          "dlq-" + mid,
						"original_message_id": mid,
						"business_event_id":   eventID,
						"source_queue_arn":    queueARN,
						"failure_reason":      messageReason(message),
						"receive_count":       receiveCount,
					}); err != nil {
						return nil, err
					}
					dlqed[mid] = true
					result["dlq_message_ids"] = append(result["dlq_message_ids"].([]any), mid)
				}
			} else {
				deleted[mid] = true
				result["deleted_message_ids"] = append(result["deleted_message_ids"].([]any), mid)
				cycleDeleted = append(cycleDeleted, mid)
			}
		}
		result["cycles"] = append(cyclesOut, map[string]any{
			"delivered": deliveredIDs,
			"failed":    cycleFailed,
			"deleted":   cycleDeleted,
		})
	}

	ledger, err := loadLedger()
	if err != nil {
		return nil, err
	}
	dlq, err := loadDLQ()
	if err != nil {
		return nil, err
	}
	result["ledger_entries"] = ledger
	result["dlq_entries"] = dlq
	return result, nil
}
